pkg/action/executor/gitops/github_manager.go

```diff
@@ -353,60 +356,131 @@ func (g *GitHandler) getTree(ref *github.Reference, fileContent []byte) (tree *g
-			return nil, fmt.Errorf("error getting branch ref %s, %v", g.baseBranch, err)
+			return nil, fmt.Errorf("error getting branch ref %s, %v", commitBranchName, err)
 		}
 	}
 
-	tree, err := g.getTree(branchRef, patchedYamlContent)
+	needsUpdate, patchedYamlContent, err := g.patchYamlContent(resName, commitBranchName, patches)
+	if err != nil {
+		return nil, err
+	}
+	if !needsUpdate {
+		if pr != nil {
+			// We have an existing PR with this action
+			// We should now wait on the existing PR to be merged
+			glog.Infof("The open pr #%d content is already aligned to desired state. "+
+				"Will now wait for the PR to be merged to complete the action.", pr.GetNumber())
+			return gitWaitData{
+				handler: g,
+				prNum:   pr.GetNumber(),
+			}, nil
+		}
+		// Its commit mode, we simply log and succeed the action
+		glog.Infof("The base branch %s is already aligned to desired state. "+
+			"Nothing to commit. Action will be considered successful.", commitBranchName)
+		return nil, nil
+	}
+
+	// If pr != nil, commitBranchRef is actually the existing PRs branch, rather
+	// then a new branch. Also, if we are here and pr != nil, that means the
+	// desired state has drifted from the existing PR. We will push a new commit
+	// on the existing PRs branch.
+	tree, err := g.createTree(commitBranchRef, patchedYamlContent)
 	if err != nil {
 		return nil, fmt.Errorf("error getting branch ref: %s, %v", g.baseBranch, err)
 	}
 
-	_, err = g.pushCommit(branchRef, tree)
+	_, err = g.pushCommit(commitBranchRef, tree)
 	if err != nil {
 		return nil, fmt.Errorf("error committing new content to branch %s, %v", g.baseBranch, err)
 	}
 
 	if g.commitMode == commitModePR {
-		pr, err := g.newPR(resName, newBranchName)
-		if err != nil {
-			return nil, err
+		if pr == nil {
+			pr, err = g.newPR(resName, resNamespace, clusterId, commitBranchName)
+			if err != nil {
+				return nil, err
+			}
 		}
-		// TODO: check if we need to validate if the pr number is set in response
-		// after the create request
+		glog.Infof("PR #%d created. "+
+			"Will now wait for the PR to be merged to complete the action.", pr.GetNumber())
 		return gitWaitData{
 			handler: g,
 			prNum:   pr.GetNumber(),
 		}, nil
 	}
+
 	return nil, nil
 }
 
+func (g *GitHandler) patchYamlContent(resName, branchName string, patches []PatchItem) (bool, []byte, error) {
+	yamlContent, err := g.getRemoteFileContent(resName, g.path, branchName)
+	if err != nil {
+		return false, nil, err
+	}
+	patchedYamlContent, err := ApplyPatch([]byte(yamlContent), patches)
+	if err != nil {
+		return false, nil, fmt.Errorf("error applying patches to file %s: patches: %v, err: %v", g.path, patches, err)
+	}
+
+	// TODO: find a better way to compare yamls. As patching can reorganise the
+	// yaml content, even if there is no change and the source yaml is not
+	// alphabetically arranged we will end up creating a commit.
+	needsUpdate := strings.Compare(yamlContent, string(patchedYamlContent)) != 0
```

Comment:
Can we just check the resource values (like cpu limit/request and mem limit/request. I am afraid there may be other noise.

Reply:
Yes, something like that is the plan as part of TODO. This is similar to what we want to address for patching too.